// @eclipse-zenoh/zenoh-ts/dist/encoding.js
//
// Copyright (c) 2024 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//
export var EncodingPredefined;
(function (EncodingPredefined) {
    EncodingPredefined[EncodingPredefined["ZENOH_BYTES"] = 0] = "ZENOH_BYTES";
    EncodingPredefined[EncodingPredefined["ZENOH_STRING"] = 1] = "ZENOH_STRING";
    EncodingPredefined[EncodingPredefined["ZENOH_SERIALIZED"] = 2] = "ZENOH_SERIALIZED";
    EncodingPredefined[EncodingPredefined["APPLICATION_OCTET_STREAM"] = 3] = "APPLICATION_OCTET_STREAM";
    EncodingPredefined[EncodingPredefined["TEXT_PLAIN"] = 4] = "TEXT_PLAIN";
    EncodingPredefined[EncodingPredefined["APPLICATION_JSON"] = 5] = "APPLICATION_JSON";
    EncodingPredefined[EncodingPredefined["TEXT_JSON"] = 6] = "TEXT_JSON";
    EncodingPredefined[EncodingPredefined["APPLICATION_CDR"] = 7] = "APPLICATION_CDR";
    EncodingPredefined[EncodingPredefined["APPLICATION_CBOR"] = 8] = "APPLICATION_CBOR";
    EncodingPredefined[EncodingPredefined["APPLICATION_YAML"] = 9] = "APPLICATION_YAML";
    EncodingPredefined[EncodingPredefined["TEXT_YAML"] = 10] = "TEXT_YAML";
    EncodingPredefined[EncodingPredefined["TEXT_JSON5"] = 11] = "TEXT_JSON5";
    EncodingPredefined[EncodingPredefined["APPLICATION_PYTHON_SERIALIZED_OBJECT"] = 12] = "APPLICATION_PYTHON_SERIALIZED_OBJECT";
    EncodingPredefined[EncodingPredefined["APPLICATION_PROTOBUF"] = 13] = "APPLICATION_PROTOBUF";
    EncodingPredefined[EncodingPredefined["APPLICATION_JAVA_SERIALIZED_OBJECT"] = 14] = "APPLICATION_JAVA_SERIALIZED_OBJECT";
    EncodingPredefined[EncodingPredefined["APPLICATION_OPENMETRICS_TEXT"] = 15] = "APPLICATION_OPENMETRICS_TEXT";
    EncodingPredefined[EncodingPredefined["IMAGE_PNG"] = 16] = "IMAGE_PNG";
    EncodingPredefined[EncodingPredefined["IMAGE_JPEG"] = 17] = "IMAGE_JPEG";
    EncodingPredefined[EncodingPredefined["IMAGE_GIF"] = 18] = "IMAGE_GIF";
    EncodingPredefined[EncodingPredefined["IMAGE_BMP"] = 19] = "IMAGE_BMP";
    EncodingPredefined[EncodingPredefined["IMAGE_WEBP"] = 20] = "IMAGE_WEBP";
    EncodingPredefined[EncodingPredefined["APPLICATION_XML"] = 21] = "APPLICATION_XML";
    EncodingPredefined[EncodingPredefined["APPLICATION_X_WWW_FORM_URLENCODED"] = 22] = "APPLICATION_X_WWW_FORM_URLENCODED";
    EncodingPredefined[EncodingPredefined["TEXT_HTML"] = 23] = "TEXT_HTML";
    EncodingPredefined[EncodingPredefined["TEXT_XML"] = 24] = "TEXT_XML";
    EncodingPredefined[EncodingPredefined["TEXT_CSS"] = 25] = "TEXT_CSS";
    EncodingPredefined[EncodingPredefined["TEXT_JAVASCRIPT"] = 26] = "TEXT_JAVASCRIPT";
    EncodingPredefined[EncodingPredefined["TEXT_MARKDOWN"] = 27] = "TEXT_MARKDOWN";
    EncodingPredefined[EncodingPredefined["TEXT_CSV"] = 28] = "TEXT_CSV";
    EncodingPredefined[EncodingPredefined["APPLICATION_SQL"] = 29] = "APPLICATION_SQL";
    EncodingPredefined[EncodingPredefined["APPLICATION_COAP_PAYLOAD"] = 30] = "APPLICATION_COAP_PAYLOAD";
    EncodingPredefined[EncodingPredefined["APPLICATION_JSON_PATCH_JSON"] = 31] = "APPLICATION_JSON_PATCH_JSON";
    EncodingPredefined[EncodingPredefined["APPLICATION_JSON_SEQ"] = 32] = "APPLICATION_JSON_SEQ";
    EncodingPredefined[EncodingPredefined["APPLICATION_JSONPATH"] = 33] = "APPLICATION_JSONPATH";
    EncodingPredefined[EncodingPredefined["APPLICATION_JWT"] = 34] = "APPLICATION_JWT";
    EncodingPredefined[EncodingPredefined["APPLICATION_MP4"] = 35] = "APPLICATION_MP4";
    EncodingPredefined[EncodingPredefined["APPLICATION_SOAP_XML"] = 36] = "APPLICATION_SOAP_XML";
    EncodingPredefined[EncodingPredefined["APPLICATION_YANG"] = 37] = "APPLICATION_YANG";
    EncodingPredefined[EncodingPredefined["AUDIO_AAC"] = 38] = "AUDIO_AAC";
    EncodingPredefined[EncodingPredefined["AUDIO_FLAC"] = 39] = "AUDIO_FLAC";
    EncodingPredefined[EncodingPredefined["AUDIO_MP4"] = 40] = "AUDIO_MP4";
    EncodingPredefined[EncodingPredefined["AUDIO_OGG"] = 41] = "AUDIO_OGG";
    EncodingPredefined[EncodingPredefined["AUDIO_VORBIS"] = 42] = "AUDIO_VORBIS";
    EncodingPredefined[EncodingPredefined["VIDEO_H261"] = 43] = "VIDEO_H261";
    EncodingPredefined[EncodingPredefined["VIDEO_H263"] = 44] = "VIDEO_H263";
    EncodingPredefined[EncodingPredefined["VIDEO_H264"] = 45] = "VIDEO_H264";
    EncodingPredefined[EncodingPredefined["VIDEO_H265"] = 46] = "VIDEO_H265";
    EncodingPredefined[EncodingPredefined["VIDEO_H266"] = 47] = "VIDEO_H266";
    EncodingPredefined[EncodingPredefined["VIDEO_MP4"] = 48] = "VIDEO_MP4";
    EncodingPredefined[EncodingPredefined["VIDEO_OGG"] = 49] = "VIDEO_OGG";
    EncodingPredefined[EncodingPredefined["VIDEO_RAW"] = 50] = "VIDEO_RAW";
    EncodingPredefined[EncodingPredefined["VIDEO_VP8"] = 51] = "VIDEO_VP8";
    EncodingPredefined[EncodingPredefined["VIDEO_VP9"] = 52] = "VIDEO_VP9";
    EncodingPredefined[EncodingPredefined["CUSTOM"] = 65535] = "CUSTOM";
})(EncodingPredefined || (EncodingPredefined = {}));
function getPredefinedEncodingString(e) {
    switch (e) {
        case EncodingPredefined.APPLICATION_OCTET_STREAM: return "application/octet-stream";
        case EncodingPredefined.APPLICATION_PYTHON_SERIALIZED_OBJECT: return "application/python-serialized-object";
        case EncodingPredefined.APPLICATION_JAVA_SERIALIZED_OBJECT: return "application/java-serialized-object";
        case EncodingPredefined.APPLICATION_OPENMETRICS_TEXT: return "application/openmetrics-text";
        case EncodingPredefined.APPLICATION_X_WWW_FORM_URLENCODED: return "application/x-www-form-urlencoded";
        case EncodingPredefined.APPLICATION_COAP_PAYLOAD: return "application/coap-payload";
        case EncodingPredefined.APPLICATION_JSON_PATCH_JSON: return "application/json-patch+json";
        case EncodingPredefined.APPLICATION_JSON_SEQ: return "application/json-seq";
        case EncodingPredefined.APPLICATION_SOAP_XML: return "application/soap+xml";
        default: return EncodingPredefined[e].toLocaleLowerCase().replaceAll('_', '/');
    }
}
function createIdToPredefinedEncodingMap() {
    let out = new Map();
    for (let e in EncodingPredefined) {
        let n = Number(e);
        if (!isNaN(n) && n != EncodingPredefined.CUSTOM) {
            out.set(n, getPredefinedEncodingString(n));
        }
    }
    return out;
}
function createPredefinedEncodingToIdMap() {
    let out = new Map();
    for (let e in EncodingPredefined) {
        let n = Number(e);
        if (!isNaN(n) && n != EncodingPredefined.CUSTOM) {
            out.set(getPredefinedEncodingString(n), n);
        }
    }
    return out;
}
/**
 * Zenoh Encoding Class
 */
export class Encoding {
    id;
    schema;
    static ID_TO_ENCODING = createIdToPredefinedEncodingMap();
    static ENCODING_TO_ID = createPredefinedEncodingToIdMap();
    static SEP = ";";
    constructor(id, schema) {
        this.id = id;
        this.schema = schema;
    }
    withSchema(input) {
        if (this.id != EncodingPredefined.CUSTOM || this.schema == undefined) {
            return new Encoding(this.id, input);
        }
        else {
            const idx = this.schema.indexOf(Encoding.SEP);
            if (idx == -1) {
                return new Encoding(this.id, this.schema + ";" + input);
            }
            else {
                return new Encoding(this.id, this.schema.substring(0, idx) + ";" + input);
            }
        }
    }
    static default() {
        return new Encoding(EncodingPredefined.ZENOH_BYTES);
    }
    toString() {
        let out = "";
        if (this.id != EncodingPredefined.CUSTOM) {
            out += Encoding.ID_TO_ENCODING.get(this.id);
        }
        if (this.schema != undefined) {
            if (this.id != EncodingPredefined.CUSTOM) {
                out += ";";
            }
            out += this.schema;
        }
        return out;
    }
    static fromString(input) {
        if (input.length == 0) {
            return new Encoding(EncodingPredefined.ZENOH_BYTES, undefined);
        }
        const idx = input.indexOf(Encoding.SEP);
        let key;
        let schema = undefined;
        if (idx == -1) {
            key = input;
        }
        else {
            key = input.substring(0, idx);
            schema = input.substring(idx + 1);
        }
        const id = Encoding.ENCODING_TO_ID.get(key) ?? EncodingPredefined.CUSTOM;
        if (id == EncodingPredefined.CUSTOM) {
            schema = schema ? key + ";" + schema : key;
        }
        return new Encoding(id, schema);
    }
    static from(input) {
        if (input instanceof Encoding) {
            return input;
        }
        else {
            return Encoding.fromString(input.toString());
        }
    }
    toIdSchema() {
        return [this.id, this.schema];
    }
    // Enum Variants
    /**
     * Constant alias for string "zenoh/bytes"
     */
    static ZENOH_BYTES = new Encoding(EncodingPredefined.ZENOH_BYTES);
    /**
     * Constant alias for string "zenoh/string"
     */
    static ZENOH_STRING = new Encoding(EncodingPredefined.ZENOH_STRING);
    /**
     * Constant alias for string "zenoh/serialized"
     */
    static ZENOH_SERIALIZED = new Encoding(EncodingPredefined.ZENOH_SERIALIZED);
    /**
     * Constant alias for string "application/octet-stream"
     */
    static APPLICATION_OCTET_STREAM = new Encoding(EncodingPredefined.APPLICATION_OCTET_STREAM);
    /**
     * Constant alias for string "text/plain"
     */
    static TEXT_PLAIN = new Encoding(EncodingPredefined.TEXT_PLAIN);
    /**
     * Constant alias for string "application/json"
     */
    static APPLICATION_JSON = new Encoding(EncodingPredefined.APPLICATION_JSON);
    /**
     * Constant alias for string "text/json"
     */
    static TEXT_JSON = new Encoding(EncodingPredefined.TEXT_JSON);
    /**
     * Constant alias for string "application/cdr"
     */
    static APPLICATION_CDR = new Encoding(EncodingPredefined.APPLICATION_CDR);
    /**
     * Constant alias for string "application/cbor"
     */
    static APPLICATION_CBOR = new Encoding(EncodingPredefined.APPLICATION_CBOR);
    /**
     * Constant alias for string "application/yaml"
     */
    static APPLICATION_YAML = new Encoding(EncodingPredefined.APPLICATION_YAML);
    /**
     * Constant alias for string "text/yaml"
     */
    static TEXT_YAML = new Encoding(EncodingPredefined.TEXT_YAML);
    /**
     * Constant alias for string "text/json5"
     */
    static TEXT_JSON5 = new Encoding(EncodingPredefined.TEXT_JSON5);
    /**
     * Constant alias for string "application/protobuf"
     */
    static APPLICATION_PROTOBUF = new Encoding(EncodingPredefined.APPLICATION_PROTOBUF);
    /**
     * Constant alias for string "application/python-serialized-object"
     */
    static APPLICATION_PYTHON_SERIALIZED_OBJECT = new Encoding(EncodingPredefined.APPLICATION_PYTHON_SERIALIZED_OBJECT);
    /**
     * Constant alias for string "application/java-serialized-object"
     */
    static APPLICATION_JAVA_SERIALIZED_OBJECT = new Encoding(EncodingPredefined.APPLICATION_JAVA_SERIALIZED_OBJECT);
    /**
     * Constant alias for string "application/openmetrics-text"
     */
    static APPLICATION_OPENMETRICS_TEXT = new Encoding(EncodingPredefined.APPLICATION_OPENMETRICS_TEXT);
    /**
     * Constant alias for string "image/png"
     */
    static IMAGE_PNG = new Encoding(EncodingPredefined.IMAGE_PNG);
    /**
     * Constant alias for string "image/jpeg"
     */
    static IMAGE_JPEG = new Encoding(EncodingPredefined.IMAGE_JPEG);
    /**
     * Constant alias for string "image/gif"
     */
    static IMAGE_GIF = new Encoding(EncodingPredefined.IMAGE_GIF);
    /**
     * Constant alias for string "image/bmp"
     */
    static IMAGE_BMP = new Encoding(EncodingPredefined.IMAGE_BMP);
    /**
     * Constant alias for string "image/webp"
     */
    static IMAGE_WEBP = new Encoding(EncodingPredefined.IMAGE_WEBP);
    /**
     * Constant alias for string "application/xml"
     */
    static APPLICATION_XML = new Encoding(EncodingPredefined.APPLICATION_XML);
    /**
     * Constant alias for string "application/x-www-form-urlencoded"
     */
    static APPLICATION_X_WWW_FORM_URLENCODED = new Encoding(EncodingPredefined.APPLICATION_X_WWW_FORM_URLENCODED);
    /**
     * Constant alias for string "text/html"
     */
    static TEXT_HTML = new Encoding(EncodingPredefined.TEXT_HTML);
    /**
     * Constant alias for string "text/xml"
     */
    static TEXT_XML = new Encoding(EncodingPredefined.TEXT_XML);
    /**
     * Constant alias for string "text/css"
     */
    static TEXT_CSS = new Encoding(EncodingPredefined.TEXT_CSS);
    /**
     * Constant alias for string "text/javascript"
     */
    static TEXT_JAVASCRIPT = new Encoding(EncodingPredefined.TEXT_JAVASCRIPT);
    /**
     * Constant alias for string "text/markdown"
     */
    static TEXT_MARKDOWN = new Encoding(EncodingPredefined.TEXT_MARKDOWN);
    /**
     * Constant alias for string "text/csv"
     */
    static TEXT_CSV = new Encoding(EncodingPredefined.TEXT_CSV);
    /**
     * Constant alias for string "application/sql"
     */
    static APPLICATION_SQL = new Encoding(EncodingPredefined.APPLICATION_SQL);
    /**
     * Constant alias for string "application/coap-payload"
     */
    static APPLICATION_COAP_PAYLOAD = new Encoding(EncodingPredefined.APPLICATION_COAP_PAYLOAD);
    /**
     * Constant alias for string "application/json-patch+json"
     */
    static APPLICATION_JSON_PATCH_JSON = new Encoding(EncodingPredefined.APPLICATION_JSON_PATCH_JSON);
    /**
     * Constant alias for string "application/json-seq"
     */
    static APPLICATION_JSON_SEQ = new Encoding(EncodingPredefined.APPLICATION_JSON_SEQ);
    /**
     * Constant alias for string "application/jsonpath"
     */
    static APPLICATION_JSONPATH = new Encoding(EncodingPredefined.APPLICATION_JSONPATH);
    /**
     * Constant alias for string "application/jwt"
     */
    static APPLICATION_JWT = new Encoding(EncodingPredefined.APPLICATION_JWT);
    /**
     * Constant alias for string "application/mp4"
     */
    static APPLICATION_MP4 = new Encoding(EncodingPredefined.APPLICATION_MP4);
    /**
     * Constant alias for string "application/soap+xml"
     */
    static APPLICATION_SOAP_XML = new Encoding(EncodingPredefined.APPLICATION_SOAP_XML);
    /**
     * Constant alias for string "application/yang"
     */
    static APPLICATION_YANG = new Encoding(EncodingPredefined.APPLICATION_YANG);
    /**
     * Constant alias for string "audio/aac"
     */
    static AUDIO_AAC = new Encoding(EncodingPredefined.AUDIO_AAC);
    /**
     * Constant alias for string "audio/flac"
     */
    static AUDIO_FLAC = new Encoding(EncodingPredefined.AUDIO_FLAC);
    /**
     * Constant alias for string "audio/mp4"
     */
    static AUDIO_MP4 = new Encoding(EncodingPredefined.AUDIO_MP4);
    /**
     * Constant alias for string "audio/ogg"
     */
    static AUDIO_OGG = new Encoding(EncodingPredefined.AUDIO_OGG);
    /**
     * Constant alias for string "audio/vorbis"
     */
    static AUDIO_VORBIS = new Encoding(EncodingPredefined.AUDIO_VORBIS);
    /**
     * Constant alias for string "video/h261"
     */
    static VIDEO_H261 = new Encoding(EncodingPredefined.VIDEO_H261);
    /**
     * Constant alias for string "video/h263"
     */
    static VIDEO_H263 = new Encoding(EncodingPredefined.VIDEO_H263);
    /**
     * Constant alias for string "video/h264"
     */
    static VIDEO_H264 = new Encoding(EncodingPredefined.VIDEO_H264);
    /**
     * Constant alias for string "video/h265"
     */
    static VIDEO_H265 = new Encoding(EncodingPredefined.VIDEO_H265);
    /**
     * Constant alias for string "video/h266"
     */
    static VIDEO_H266 = new Encoding(EncodingPredefined.VIDEO_H266);
    /**
     * Constant alias for string "video/mp4"
     */
    static VIDEO_MP4 = new Encoding(EncodingPredefined.VIDEO_MP4);
    /**
     * Constant alias for string "video/ogg"
     */
    static VIDEO_OGG = new Encoding(EncodingPredefined.VIDEO_OGG);
    /**
     * Constant alias for string "video/raw"
     */
    static VIDEO_RAW = new Encoding(EncodingPredefined.VIDEO_RAW);
    /**
     * Constant alias for string "video/vp8"
     */
    static VIDEO_VP8 = new Encoding(EncodingPredefined.VIDEO_VP8);
    /**
     * Constant alias for string "video/vp9"
     */
    static VIDEO_VP9 = new Encoding(EncodingPredefined.VIDEO_VP9);
}
