// @eclipse-zenoh/zenoh-ts/dist/matching.js
//
// Copyright (c) 2025 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//
/**
 * A class that indicates if there exist Subscribers matching the Publisher's key expression or Queryables matching Querier's key expression and target.
 */
export class MatchingStatus {
    matching_;
    /**
     * @internal
     */
    constructor(matching_) {
        this.matching_ = matching_;
    }
    /**
     *
     * @returns ``True`` if there exist matching Zenoh entities, ``false`` otherwise.
     */
    matching() {
        return this.matching_;
    }
}
/**
 * A Zenoh matching listener that sends notifications when the `MatchingStatus` of a `Publisher` or a `Querier` changes.
 */
export class MatchingListener {
    session;
    id;
    receiver_;
    /**
     * @ignore
     */
    async [Symbol.asyncDispose]() {
        await this.undeclare();
    }
    /**
     * @ignore
     */
    constructor(session, id, receiver_) {
        this.session = session;
        this.id = id;
        this.receiver_ = receiver_;
    }
    /**
     * returns a message status receiver for non-callback matching listener, undefined otherwise.
     *
     * @returns ChannelReceiver<MatchingStatus> | undefined
     */
    receiver() {
        return this.receiver_;
    }
    /**
     * Undeclares a subscriber on the session
     *
     */
    async undeclare() {
        await this.session.undeclareMatchingListener(this.id);
    }
}
