// @eclipse-zenoh/zenoh-ts/dist/message.js
//
// Copyright (c) 2025 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//
import { ZD } from "./ext/index.js";
import { Encoding } from "./encoding.js";
import { KeyExpr } from "./key_expr.js";
import { Locality, Reliability, CongestionControl, Priority, SampleKind, ReplyKeyExpr } from "./enums.js";
import { Timestamp } from "./timestamp.js";
import { ZenohId } from "./zid.js";
import { Sample } from "./sample.js";
import { Parameters, QueryInner, Reply, ReplyError } from "./query.js";
import { ZBytes } from "./z_bytes.js";
import { SessionInfo } from "./session.js";
function sampleKindFromUint8(val) {
    switch (val) {
        case SampleKind.PUT: return SampleKind.PUT;
        case SampleKind.DELETE: return SampleKind.DELETE;
        default: {
            console.warn(`Unsupported SampleKind value ${val}`);
            return SampleKind.PUT;
        }
    }
}
function priorityFromUint8(val) {
    switch (val) {
        case Priority.REAL_TIME: return Priority.REAL_TIME;
        case Priority.INTERACTIVE_HIGH: return Priority.INTERACTIVE_HIGH;
        case Priority.INTERACTIVE_LOW: return Priority.INTERACTIVE_LOW;
        case Priority.DATA_HIGH: return Priority.DATA_HIGH;
        case Priority.DATA: return Priority.DATA;
        case Priority.DATA_LOW: return Priority.DATA_LOW;
        case Priority.BACKGROUND: return Priority.BACKGROUND;
        default: {
            console.warn(`Unsupported Priority value ${val}`);
            return Priority.DEFAULT;
        }
    }
}
function congestionControlFromUint8(val) {
    switch (val) {
        case CongestionControl.DROP: return CongestionControl.DROP;
        case CongestionControl.BLOCK: return CongestionControl.BLOCK;
        default: {
            console.warn(`Unsupported CongestionControl value ${val}`);
            return CongestionControl.DEFAULT_PUSH;
        }
    }
}
function reliabilityFromUint8(val) {
    switch (val) {
        case Reliability.BEST_EFFORT: return Reliability.BEST_EFFORT;
        case Reliability.RELIABLE: return Reliability.RELIABLE;
        default: {
            console.warn(`Unsupported Reliability value ${val}`);
            return Reliability.DEFAULT;
        }
    }
}
function localityFromUint8(val) {
    switch (val) {
        case Locality.SESSION_LOCAL: return Locality.SESSION_LOCAL;
        case Locality.REMOTE: return Locality.REMOTE;
        case Locality.ANY: return Locality.ANY;
        default: {
            console.warn(`Unsupported Locality value ${val}`);
            return Locality.DEFAULT;
        }
    }
}
export class Qos {
    priority;
    congestionControl;
    express;
    reliability;
    locality;
    constructor(priority, congestionControl, express, reliability, locality) {
        this.priority = priority;
        this.congestionControl = congestionControl;
        this.express = express;
        this.reliability = reliability;
        this.locality = locality;
    }
}
function qosToUint8(qos) {
    // llrecppp
    let e = qos.express ? 1 : 0;
    return qos.priority | (qos.congestionControl << 3) | (e << 4) | (qos.reliability << 5) | (qos.locality << 6);
}
function qosFromUint8(val) {
    let p = val & 0b111;
    let c = (val >> 3) & 0b1;
    let e = (val >> 4) & 0b1;
    let r = (val >> 5) & 0b1;
    let l = (val >> 6) & 0b11;
    return new Qos(priorityFromUint8(p), congestionControlFromUint8(c), e != 0, reliabilityFromUint8(r), localityFromUint8(l));
}
function replyKeyExprFromUint8(val) {
    switch (val) {
        case ReplyKeyExpr.ANY: return ReplyKeyExpr.ANY;
        case ReplyKeyExpr.MATCHING_QUERY: return ReplyKeyExpr.MATCHING_QUERY;
        default: {
            console.warn(`Unsupported ReplyKeyExpr value ${val}`);
            return ReplyKeyExpr.DEFAULT;
        }
    }
}
function serializeEncoding(e, serializer) {
    let [id, schema] = e.toIdSchema();
    serializer.serializeNumberUint16(id);
    serializer.serializeString(schema ?? "");
}
function serializeOptEncoding(e, serializer) {
    if (e == undefined) {
        serializer.serializeBoolean(false);
    }
    else {
        serializer.serializeBoolean(true);
        serializeEncoding(e, serializer);
    }
}
function deserializeEncoding(deserializer) {
    let id = deserializer.deserializeNumberUint16();
    let schema = deserializer.deserializeString();
    if (schema.length == 0) {
        schema = undefined;
    }
    return new Encoding(id, schema);
}
function deserializeOptEncoding(deserializer) {
    if (deserializer.deserializeBoolean()) {
        return deserializeEncoding(deserializer);
    }
    else {
        return undefined;
    }
}
export class QuerySettings {
    target;
    consolidation;
    replyKeyExpr;
    constructor(target, consolidation, replyKeyExpr) {
        this.target = target;
        this.consolidation = consolidation;
        this.replyKeyExpr = replyKeyExpr;
    }
}
function querySettingsToUint8(qs) {
    // rcctt
    return qs.target | (qs.consolidation << 2) | (qs.replyKeyExpr << 4);
}
function deserializeZenohId(deserializer) {
    return new ZenohId(deserializer.deserializeUint8Array());
}
function serializeZid(zid, serializer) {
    serializer.serializeUint8Array(zid.toLeBytes());
}
function serializeTimestamp(timestamp, serializer) {
    serializer.serializeBigintUint64(timestamp.getNtp64());
    serializeZid(timestamp.getId(), serializer);
}
function serializeOptTimestamp(timestamp, serializer) {
    if (timestamp == undefined) {
        serializer.serializeBoolean(false);
    }
    else {
        serializer.serializeBoolean(true);
        serializeTimestamp(timestamp, serializer);
    }
}
function deserializeTimestamp(deserializer) {
    let ntp64 = deserializer.deserializeBigintUint64();
    let zid = deserializeZenohId(deserializer);
    return new Timestamp(zid, ntp64);
}
function deserializeOptTimestamp(deserializer) {
    if (deserializer.deserializeBoolean()) {
        return deserializeTimestamp(deserializer);
    }
    else {
        return undefined;
    }
}
function serializeOptZBytes(a, serializer) {
    if (a == undefined) {
        serializer.serialize(false);
    }
    else {
        serializer.serialize(true);
        serializer.serializeUint8Array(a.toBytes());
    }
}
function deserializeOptZBytes(deserializer) {
    if (deserializer.deserializeBoolean()) {
        return new ZBytes(deserializer.deserializeUint8Array());
    }
    else {
        return undefined;
    }
}
function deserializeSample(deserializer) {
    let keyexpr = new KeyExpr(deserializer.deserializeString());
    let payload = new ZBytes(deserializer.deserializeUint8Array());
    let kind = sampleKindFromUint8(deserializer.deserializeNumberUint8());
    let encoding = deserializeEncoding(deserializer);
    let attachment = deserializeOptZBytes(deserializer);
    let timestamp = deserializeOptTimestamp(deserializer);
    let qos = qosFromUint8(deserializer.deserializeNumberUint8());
    return new Sample(keyexpr, payload, kind, encoding, attachment, timestamp, qos.priority, qos.congestionControl, qos.express);
}
function deserializeReply(deserializer) {
    if (deserializer.deserializeBoolean()) {
        return new Reply(deserializeSample(deserializer));
    }
    else {
        let payload = new ZBytes(deserializer.deserializeUint8Array());
        let encoding = deserializeEncoding(deserializer);
        return new Reply(new ReplyError(payload, encoding));
    }
}
function deserializeQueryInner(deserializer) {
    let queryId = deserializer.deserializeNumberUint32();
    let keyexpr = new KeyExpr(deserializer.deserializeString());
    let params = new Parameters(deserializer.deserializeString());
    let payload = deserializeOptZBytes(deserializer);
    let encoding = deserializeOptEncoding(deserializer);
    let attachment = deserializeOptZBytes(deserializer);
    let replyKeyExpr = replyKeyExprFromUint8(deserializer.deserializeNumberUint8());
    return new QueryInner(queryId, keyexpr, params, payload, encoding, attachment, replyKeyExpr);
}
export var OutRemoteMessageId;
(function (OutRemoteMessageId) {
    OutRemoteMessageId[OutRemoteMessageId["DeclarePublisher"] = 0] = "DeclarePublisher";
    OutRemoteMessageId[OutRemoteMessageId["UndeclarePublisher"] = 1] = "UndeclarePublisher";
    OutRemoteMessageId[OutRemoteMessageId["DeclareSubscriber"] = 2] = "DeclareSubscriber";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareSubscriber"] = 3] = "UndeclareSubscriber";
    OutRemoteMessageId[OutRemoteMessageId["DeclareQueryable"] = 4] = "DeclareQueryable";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareQueryable"] = 5] = "UndeclareQueryable";
    OutRemoteMessageId[OutRemoteMessageId["DeclareQuerier"] = 6] = "DeclareQuerier";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareQuerier"] = 7] = "UndeclareQuerier";
    OutRemoteMessageId[OutRemoteMessageId["DeclareLivelinessToken"] = 8] = "DeclareLivelinessToken";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareLivelinessToken"] = 9] = "UndeclareLivelinessToken";
    OutRemoteMessageId[OutRemoteMessageId["DeclareLivelinessSubscriber"] = 10] = "DeclareLivelinessSubscriber";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareLivelinessSubscriber"] = 11] = "UndeclareLivelinessSubscriber";
    OutRemoteMessageId[OutRemoteMessageId["GetSessionInfo"] = 12] = "GetSessionInfo";
    OutRemoteMessageId[OutRemoteMessageId["GetTimestamp"] = 13] = "GetTimestamp";
    OutRemoteMessageId[OutRemoteMessageId["Put"] = 14] = "Put";
    OutRemoteMessageId[OutRemoteMessageId["Delete"] = 15] = "Delete";
    OutRemoteMessageId[OutRemoteMessageId["PublisherPut"] = 16] = "PublisherPut";
    OutRemoteMessageId[OutRemoteMessageId["PublisherDelete"] = 17] = "PublisherDelete";
    OutRemoteMessageId[OutRemoteMessageId["Get"] = 18] = "Get";
    OutRemoteMessageId[OutRemoteMessageId["QuerierGet"] = 19] = "QuerierGet";
    OutRemoteMessageId[OutRemoteMessageId["LivelinessGet"] = 20] = "LivelinessGet";
    OutRemoteMessageId[OutRemoteMessageId["ReplyOk"] = 21] = "ReplyOk";
    OutRemoteMessageId[OutRemoteMessageId["ReplyDel"] = 22] = "ReplyDel";
    OutRemoteMessageId[OutRemoteMessageId["ReplyErr"] = 23] = "ReplyErr";
    OutRemoteMessageId[OutRemoteMessageId["QueryResponseFinal"] = 24] = "QueryResponseFinal";
    OutRemoteMessageId[OutRemoteMessageId["Ping"] = 25] = "Ping";
    OutRemoteMessageId[OutRemoteMessageId["PublisherDeclareMatchingListener"] = 26] = "PublisherDeclareMatchingListener";
    OutRemoteMessageId[OutRemoteMessageId["UndeclareMatchingListener"] = 27] = "UndeclareMatchingListener";
    OutRemoteMessageId[OutRemoteMessageId["PublisherGetMatchingStatus"] = 28] = "PublisherGetMatchingStatus";
    OutRemoteMessageId[OutRemoteMessageId["QuerierDeclareMatchingListener"] = 29] = "QuerierDeclareMatchingListener";
    OutRemoteMessageId[OutRemoteMessageId["QuerierGetMatchingStatus"] = 30] = "QuerierGetMatchingStatus";
})(OutRemoteMessageId || (OutRemoteMessageId = {}));
export class DeclarePublisher {
    id;
    properties;
    outMessageId = OutRemoteMessageId.DeclarePublisher;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializeEncoding(this.properties.encoding, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.properties.qos));
    }
}
export class UndeclarePublisher {
    id;
    outMessageId = OutRemoteMessageId.UndeclarePublisher;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class DeclareSubscriber {
    id;
    properties;
    outMessageId = OutRemoteMessageId.DeclareSubscriber;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeNumberUint8(this.properties.allowedOrigin);
    }
}
export class UndeclareSubscriber {
    id;
    outMessageId = OutRemoteMessageId.UndeclareSubscriber;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class DeclareQueryable {
    id;
    properties;
    outMessageId = OutRemoteMessageId.DeclareQueryable;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeBoolean(this.properties.complete);
        serializer.serializeNumberUint8(this.properties.allowedOrigin);
    }
}
export class UndeclareQueryable {
    id;
    outMessageId = OutRemoteMessageId.UndeclareQueryable;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class DeclareQuerier {
    id;
    properties;
    outMessageId = OutRemoteMessageId.DeclareQuerier;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeNumberUint8(qosToUint8(this.properties.qos));
        serializer.serializeNumberUint8(querySettingsToUint8(this.properties.querySettings));
        serializer.serializeNumberUint32(this.properties.timeoutMs);
    }
}
export class UndeclareQuerier {
    id;
    outMessageId = OutRemoteMessageId.UndeclareQuerier;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class DeclareLivelinessToken {
    id;
    keyexpr;
    outMessageId = OutRemoteMessageId.DeclareLivelinessToken;
    constructor(id, keyexpr) {
        this.id = id;
        this.keyexpr = keyexpr;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.keyexpr.toString());
    }
}
export class UndeclareLivelinessToken {
    id;
    outMessageId = OutRemoteMessageId.UndeclareLivelinessToken;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class DeclareLivelinessSubscriber {
    id;
    properties;
    outMessageId = OutRemoteMessageId.DeclareLivelinessSubscriber;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeBoolean(this.properties.history);
    }
}
export class UndeclareLivelinessSubscriber {
    id;
    outMessageId = OutRemoteMessageId.UndeclareLivelinessSubscriber;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class GetSessionInfo {
    outMessageId = OutRemoteMessageId.GetSessionInfo;
    constructor() { }
    serializeWithZSerializer(_serializer) { }
}
export class GetTimestamp {
    outMessageId = OutRemoteMessageId.GetTimestamp;
    constructor() { }
    serializeWithZSerializer(_serializer) { }
}
export class Put {
    keyexpr;
    payload;
    encoding;
    attachment;
    timestamp;
    qos;
    outMessageId = OutRemoteMessageId.Put;
    constructor(keyexpr, payload, encoding, attachment, timestamp, qos) {
        this.keyexpr = keyexpr;
        this.payload = payload;
        this.encoding = encoding;
        this.attachment = attachment;
        this.timestamp = timestamp;
        this.qos = qos;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeString(this.keyexpr.toString());
        serializer.serializeUint8Array(this.payload.toBytes());
        serializeEncoding(this.encoding, serializer);
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.qos));
    }
}
export class Delete {
    keyexpr;
    attachment;
    timestamp;
    qos;
    outMessageId = OutRemoteMessageId.Delete;
    constructor(keyexpr, attachment, timestamp, qos) {
        this.keyexpr = keyexpr;
        this.attachment = attachment;
        this.timestamp = timestamp;
        this.qos = qos;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeString(this.keyexpr.toString());
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.qos));
    }
}
export class PublisherPut {
    publisherId;
    payload;
    encoding;
    attachment;
    timestamp;
    outMessageId = OutRemoteMessageId.PublisherPut;
    constructor(publisherId, payload, encoding, attachment, timestamp) {
        this.publisherId = publisherId;
        this.payload = payload;
        this.encoding = encoding;
        this.attachment = attachment;
        this.timestamp = timestamp;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.publisherId);
        serializer.serializeUint8Array(this.payload.toBytes());
        serializeOptEncoding(this.encoding, serializer);
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
    }
}
export class PublisherDelete {
    publisherId;
    attachment;
    timestamp;
    outMessageId = OutRemoteMessageId.PublisherDelete;
    constructor(publisherId, attachment, timestamp) {
        this.publisherId = publisherId;
        this.attachment = attachment;
        this.timestamp = timestamp;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.publisherId);
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
    }
}
export class Get {
    id;
    properties;
    outMessageId = OutRemoteMessageId.Get;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeString(this.properties.parameters);
        serializeOptZBytes(this.properties.payload, serializer);
        serializeOptEncoding(this.properties.encoding, serializer);
        serializeOptZBytes(this.properties.attachment, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.properties.qos));
        serializer.serializeNumberUint8(querySettingsToUint8(this.properties.querySettings));
        serializer.serializeNumberUint32(this.properties.timeoutMs);
    }
}
export class QuerierGet {
    id;
    properties;
    outMessageId = OutRemoteMessageId.QuerierGet;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeNumberUint32(this.properties.querierId);
        serializer.serializeString(this.properties.parameters);
        serializeOptZBytes(this.properties.payload, serializer);
        serializeOptEncoding(this.properties.encoding, serializer);
        serializeOptZBytes(this.properties.attachment, serializer);
    }
}
export class LivelinessGet {
    id;
    properties;
    outMessageId = OutRemoteMessageId.LivelinessGet;
    constructor(id, properties) {
        this.id = id;
        this.properties = properties;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeString(this.properties.keyexpr.toString());
        serializer.serializeNumberUint32(this.properties.timeoutMs);
    }
}
export class ReplyOk {
    queryId;
    keyexpr;
    payload;
    encoding;
    attachment;
    timestamp;
    qos;
    outMessageId = OutRemoteMessageId.ReplyOk;
    constructor(queryId, keyexpr, payload, encoding, attachment, timestamp, qos) {
        this.queryId = queryId;
        this.keyexpr = keyexpr;
        this.payload = payload;
        this.encoding = encoding;
        this.attachment = attachment;
        this.timestamp = timestamp;
        this.qos = qos;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.queryId);
        serializer.serializeString(this.keyexpr.toString());
        serializer.serializeUint8Array(this.payload.toBytes());
        serializeEncoding(this.encoding, serializer);
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.qos));
    }
}
export class ReplyDel {
    queryId;
    keyexpr;
    attachment;
    timestamp;
    qos;
    outMessageId = OutRemoteMessageId.ReplyDel;
    constructor(queryId, keyexpr, attachment, timestamp, qos) {
        this.queryId = queryId;
        this.keyexpr = keyexpr;
        this.attachment = attachment;
        this.timestamp = timestamp;
        this.qos = qos;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.queryId);
        serializer.serializeString(this.keyexpr.toString());
        serializeOptZBytes(this.attachment, serializer);
        serializeOptTimestamp(this.timestamp, serializer);
        serializer.serializeNumberUint8(qosToUint8(this.qos));
    }
}
export class ReplyErr {
    queryId;
    payload;
    encoding;
    outMessageId = OutRemoteMessageId.ReplyErr;
    constructor(queryId, payload, encoding) {
        this.queryId = queryId;
        this.payload = payload;
        this.encoding = encoding;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.queryId);
        serializer.serializeUint8Array(this.payload.toBytes());
        serializeEncoding(this.encoding, serializer);
    }
}
export class QueryResponseFinal {
    queryId;
    outMessageId = OutRemoteMessageId.QueryResponseFinal;
    inMessageId = InRemoteMessageId.QueryResponseFinal;
    constructor(queryId) {
        this.queryId = queryId;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.queryId);
    }
    static deserialize(deserializer) {
        let queryId = deserializer.deserializeNumberUint32();
        return new QueryResponseFinal(queryId);
    }
}
export class Ping {
    outMessageId = OutRemoteMessageId.Ping;
    constructor() { }
    serializeWithZSerializer(_serializer) { }
}
export class PublisherDeclareMatchingListener {
    id;
    publisherId;
    outMessageId = OutRemoteMessageId.PublisherDeclareMatchingListener;
    constructor(id, publisherId) {
        this.id = id;
        this.publisherId = publisherId;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeNumberUint32(this.publisherId);
    }
}
export class UndeclareMatchingListener {
    id;
    outMessageId = OutRemoteMessageId.UndeclareMatchingListener;
    constructor(id) {
        this.id = id;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
    }
}
export class PublisherGetMatchingStatus {
    publisherId;
    outMessageId = OutRemoteMessageId.PublisherGetMatchingStatus;
    constructor(publisherId) {
        this.publisherId = publisherId;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.publisherId);
    }
}
export class QuerierDeclareMatchingListener {
    id;
    querierId;
    outMessageId = OutRemoteMessageId.QuerierDeclareMatchingListener;
    constructor(id, querierId) {
        this.id = id;
        this.querierId = querierId;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.id);
        serializer.serializeNumberUint32(this.querierId);
    }
}
export class QuerierGetMatchingStatus {
    querierId;
    outMessageId = OutRemoteMessageId.QuerierGetMatchingStatus;
    constructor(querierId) {
        this.querierId = querierId;
    }
    serializeWithZSerializer(serializer) {
        serializer.serializeNumberUint32(this.querierId);
    }
}
export var InRemoteMessageId;
(function (InRemoteMessageId) {
    InRemoteMessageId[InRemoteMessageId["ResponsePing"] = 0] = "ResponsePing";
    InRemoteMessageId[InRemoteMessageId["ResponseOk"] = 1] = "ResponseOk";
    InRemoteMessageId[InRemoteMessageId["ResponseError"] = 2] = "ResponseError";
    InRemoteMessageId[InRemoteMessageId["ResponseTimestamp"] = 3] = "ResponseTimestamp";
    InRemoteMessageId[InRemoteMessageId["ResponseSessionInfo"] = 4] = "ResponseSessionInfo";
    InRemoteMessageId[InRemoteMessageId["InSample"] = 5] = "InSample";
    InRemoteMessageId[InRemoteMessageId["InQuery"] = 6] = "InQuery";
    InRemoteMessageId[InRemoteMessageId["InReply"] = 7] = "InReply";
    InRemoteMessageId[InRemoteMessageId["QueryResponseFinal"] = 8] = "QueryResponseFinal";
    InRemoteMessageId[InRemoteMessageId["ResponseMatchingStatus"] = 9] = "ResponseMatchingStatus";
    InRemoteMessageId[InRemoteMessageId["MatchingStatusUpdate"] = 10] = "MatchingStatusUpdate";
})(InRemoteMessageId || (InRemoteMessageId = {}));
export class ResponsePing {
    uuid;
    inMessageId = InRemoteMessageId.ResponsePing;
    constructor(uuid) {
        this.uuid = uuid;
    }
    static deserialize(deserializer) {
        let uuid = deserializer.deserializeString();
        return new ResponsePing(uuid);
    }
}
export class ResponseOk {
    contentId;
    inMessageId = InRemoteMessageId.ResponseOk;
    constructor(contentId) {
        this.contentId = contentId;
    }
    static deserialize(deserializer) {
        let contentId = deserializer.deserializeNumberUint8();
        return new ResponseOk(contentId);
    }
}
export class ResponseError {
    error;
    inMessageId = InRemoteMessageId.ResponseError;
    constructor(error) {
        this.error = error;
    }
    static deserialize(deserializer) {
        let error = deserializer.deserializeString();
        return new ResponseError(error);
    }
}
export class ResponseTimestamp {
    timestamp;
    inMessageId = InRemoteMessageId.ResponseTimestamp;
    constructor(timestamp) {
        this.timestamp = timestamp;
    }
    static deserialize(deserializer) {
        return new ResponseTimestamp(deserializeTimestamp(deserializer));
    }
}
export class ResponseSessionInfo {
    info;
    inMessageId = InRemoteMessageId.ResponseSessionInfo;
    constructor(info) {
        this.info = info;
    }
    static deserialize(deserializer) {
        let zid = deserializeZenohId(deserializer);
        let dt = ZD.array(ZD.objectStatic(deserializeZenohId));
        let routers = deserializer.deserialize(dt);
        let peers = deserializer.deserialize(dt);
        return new ResponseSessionInfo(new SessionInfo(zid, peers, routers));
    }
}
export class InSample {
    subscriberId;
    sample;
    inMessageId = InRemoteMessageId.InSample;
    constructor(subscriberId, sample) {
        this.subscriberId = subscriberId;
        this.sample = sample;
    }
    static deserialize(deserializer) {
        let subscriberId = deserializer.deserializeNumberUint32();
        let sample = deserializeSample(deserializer);
        return new InSample(subscriberId, sample);
    }
}
export class InQuery {
    queryableId;
    query;
    inMessageId = InRemoteMessageId.InQuery;
    constructor(queryableId, query) {
        this.queryableId = queryableId;
        this.query = query;
    }
    static deserialize(deserializer) {
        let queryableId = deserializer.deserializeNumberUint32();
        let query = deserializeQueryInner(deserializer);
        return new InQuery(queryableId, query);
    }
}
export class InReply {
    queryId;
    reply;
    inMessageId = InRemoteMessageId.InReply;
    constructor(queryId, reply) {
        this.queryId = queryId;
        this.reply = reply;
    }
    static deserialize(deserializer) {
        let queryId = deserializer.deserializeNumberUint32();
        let reply = deserializeReply(deserializer);
        return new InReply(queryId, reply);
    }
}
export class ResponseMatchingStatus {
    matching;
    inMessageId = InRemoteMessageId.ResponseMatchingStatus;
    constructor(matching) {
        this.matching = matching;
    }
    static deserialize(deserializer) {
        let matching = deserializer.deserializeBoolean();
        return new ResponseMatchingStatus(matching);
    }
}
export class MatchingStatusUpdate {
    matchingListenerId;
    matching;
    inMessageId = InRemoteMessageId.MatchingStatusUpdate;
    constructor(matchingListenerId, matching) {
        this.matchingListenerId = matchingListenerId;
        this.matching = matching;
    }
    static deserialize(deserializer) {
        let matchingListenerId = deserializer.deserializeNumberUint32();
        let matching = deserializer.deserializeBoolean();
        return new MatchingStatusUpdate(matchingListenerId, matching);
    }
}
const ID_PRESENCE_FLAG = 0b10000000;
const MESSAGE_ID_MASK = 0b01111111;
export function deserializeHeader(deserializer) {
    let messageId = deserializer.deserializeNumberUint8();
    if ((messageId & ID_PRESENCE_FLAG) == 0) {
        return [messageId, undefined];
    }
    else {
        messageId = messageId & MESSAGE_ID_MASK;
        const sequenceId = deserializer.deserializeNumberUint32();
        return [messageId, sequenceId];
    }
}
export function serializeHeader(header, serializer) {
    if (header[1] == undefined) {
        serializer.serializeNumberUint8(header[0]);
    }
    else {
        serializer.serializeNumberUint8(header[0] | ID_PRESENCE_FLAG);
        serializer.serializeNumberUint32(header[1]);
    }
}
