// @eclipse-zenoh/zenoh-ts/dist/sample.js
//
// Copyright (c) 2024 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//
export class Sample {
    keyexpr_;
    payload_;
    kind_;
    encoding_;
    attachment_;
    timestamp_;
    priority_;
    congestionControl_;
    express_;
    /**
     * @internal
     */
    constructor(keyexpr_, payload_, kind_, encoding_, attachment_, timestamp_, priority_, congestionControl_, express_) {
        this.keyexpr_ = keyexpr_;
        this.payload_ = payload_;
        this.kind_ = kind_;
        this.encoding_ = encoding_;
        this.attachment_ = attachment_;
        this.timestamp_ = timestamp_;
        this.priority_ = priority_;
        this.congestionControl_ = congestionControl_;
        this.express_ = express_;
    }
    keyexpr() {
        return this.keyexpr_;
    }
    payload() {
        return this.payload_;
    }
    kind() {
        return this.kind_;
    }
    encoding() {
        return this.encoding_;
    }
    timestamp() {
        return this.timestamp_;
    }
    congestionControl() {
        return this.congestionControl_;
    }
    priority() {
        return this.priority_;
    }
    express() {
        return this.express_;
    }
    attachment() {
        return this.attachment_;
    }
}
