// channel-ts/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var channel_1 = require("./channel");
Object.defineProperty(exports, "SimpleChannel", { enumerable: true, get: function () { return channel_1.SimpleChannel; } });
Object.defineProperty(exports, "ChannelState", { enumerable: true, get: function () { return channel_1.State; } });
Object.defineProperty(exports, "MultiReceiverChannel", { enumerable: true, get: function () { return channel_1.MultiReceiverChannel; } });
Object.defineProperty(exports, "TryReceivedKind", { enumerable: true, get: function () { return channel_1.TryReceivedKind; } });
var observe_1 = require("./observe");
Object.defineProperty(exports, "observe", { enumerable: true, get: function () { return observe_1.observe; } });
var mutex_1 = require("./mutex");
Object.defineProperty(exports, "Mutex", { enumerable: true, get: function () { return mutex_1.Mutex; } });
Object.defineProperty(exports, "Guard", { enumerable: true, get: function () { return mutex_1.Guard; } });
